src/main.rs

```diff
@@ -34,6 +46,327 @@ pub mod test_harness;
+    // We assume that this event is emitted by the orderbook.
+    OrderExpired(Order),
+
+    // The orderbook registered that somebody wants to take a specific order.
+    // We assume that this event is emitted by the orderbook.
+    OrderTakeRequest(Order), // Emitted by orderbook
+
+    // Message from the network that is emitted at the end of the announce protocol when the
+    // swap is finalized.
+    SwapFinalized(LocalSwapId, RemoteData),
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Action {
+    PublishNewOrder(Order),
+    StartSwapExecution(LocalSwapId, RemoteData),
+}
+
+// Bundles the state of the application
+pub struct Maker {
+    btc_balance: bitcoin::Amount,
+    dai_balance: dai::Amount,
+    btc_fee: bitcoin::Amount,
+    dai_fee: dai::Amount,
+    btc_locked_funds: bitcoin::Amount,
+    dai_locked_funds: dai::Amount,
+    btc_max_sell_amount: bitcoin::Amount,
+    dai_max_sell_amount: dai::Amount,
+    rate: MidMarketRate,
+    spread: Spread,
+
+    active_orders: u32,
+}
+
+impl Maker {
+    // TODO: Proper constructor
+    pub fn new(intial_bicoin_balance: crate::bitcoin::Amount, initial_rate: MidMarketRate) -> Self {
+        let ZERO_DAI = dai::Amount::from_dai_trunc(0.0).unwrap();
+        let ZERO_BTC = crate::bitcoin::Amount::from_btc(0.0).unwrap();
+
+        Maker {
+            btc_balance: intial_bicoin_balance,
+            dai_balance: ZERO_DAI.clone(),
+            btc_fee: ZERO_BTC,
+            dai_fee: ZERO_DAI.clone(),
+            btc_locked_funds: ZERO_BTC,
+            dai_locked_funds: ZERO_DAI.clone(),
+            btc_max_sell_amount: ZERO_BTC,
+            dai_max_sell_amount: ZERO_DAI,
+            rate: initial_rate,
+            // 300 ^= 3.00%
+            spread: Spread::new(300).expect("default spread works"),
+            active_orders: 0,
+        }
+    }
+
+    pub fn handle_event(&mut self, event: NetworkEvent) -> Option<Action> {
+        match event {
+            NetworkEvent::OrderExpired(order) => {
+                // TODO: Bookkeeping, decisionmaking if new order
+
+                // TODO: creating a new order should not take the wallet / book
+                // let new_order = new_dai_bitcoin_order();
+                let new_order = Order {
+                    expiration_timestamp: Utc::now(),
+                };
+
+                Some(Action::PublishNewOrder(order))
+            }
+            NetworkEvent::OrderTakeRequest(order) => {
+                // TODO: Decisionmaking if we actually allow taking the order
+
+                // TODO: Action that indicates if we accepted or declined the order to be emitted, so we can react in event loop
+
+                unimplemented!()
+            }
+            NetworkEvent::SwapFinalized(local_swap_id, remote_date) => {
+                // TODO: Bookkeeping, lock up funds
+                Some(Action::StartSwapExecution(local_swap_id, remote_date))
+            }
+            _ => None,
+        }
+    }
```

Comment:
You might want to decouple the `Maker` from the `NetworkEvent` by simply providing functions that are equivalent to the individual variants of the enum, i.e in the loop, do:

```rust
match network_event {
	NetworkEvent::OrderExpired(order) => maker.handle_order_expired(order)
}
```
That way, the maker doesn't need to care about how he learns that the order is expired (i.e. from the network).

We already do the same thing with the action. We match on it an call functions on the appropriate other components and not pass the action to the component :)

Reply:
Thinking more about this led me to the following - I think more consistent - design:

1. Give the maker "command" like functions similar to what is outlined above:

`maker.expire_order()`

2. Make each of those functions return an `Event` about what happened as a result (and not an action):

`fn expire(&mut self) -> MakerEvent`

```
enum MakerEvent {
    OrderCreated
}
```

The reasoning is that if the Maker is supposed to be in charge of everything, then the functions should be commands. Those can then potentially fail by returning a `Result` for example. Sounds useful for `Maker.take_order`.
Generally, it decouples unrelated functionality from each other, which should simplify the type signature and the implementation.

Regarding Event vs Action:

After we have executed a command, the state already changed inside the maker.
So the fact that a new order was created and the maker did the accounting for it is not an action but describes something in the past. In other words: an event.

The fact that we should publish an order after it has been created is actually not that important.
Within the test, it doesn't make sense for example because the test can't actually follow the action :)

Reply:
> Those can then potentially fail by returning a Result for example. Sounds useful for Maker.take_order.

That is something we will need - otherwise a "generic" action is needed that is either "actually take and execute" or "don't take and report back to counterparty. 

I will give your suggestion a try and see how it feels. Even though I like the idea of `Event -> Action` (an event on the network leads to an action) that does not really go together well and I agree that the Maker would emit events. (We actually had this in one of the iterations, but not exactly like outlined :)